proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,237 @@
+from the actual distribution of rewards at the epoch boundary by dividing the
+process into two distinct phases: (1) rewards calculation, and (2) rewards
+distribution.
+
+
+### Rewards Calculation
+
+In contrast to the current method of computing epoch rewards, which involves
+directly computing rewards at the first slot of the epoch boundary in the bank
+runtime, the new approach involves moving the reward computation to a
+background service known as the `EpochRewardCalculationService`. This service
+starts automatically when a validator starts and listens to a designated
+channel for any incoming rewards calculation requests, performing the necessary
+rewards calculations accordingly.
+
+When a block first crosses the epoch boundary, its bank will send a request to
+the `EpochRewardCalculationService` to start the process of epoch reward
+computation. This marks the start of the reward computation phase.
+
+However, multiple rewards computation requests can be generated when there are
+forks at the epoch boundaries.
+
+The following diagram illustrates a fork happening at the epoch boundary `N`.
+
+```
+N-1 -- N -- N+1
+     \
+      \
+        N+2
+```
+
+Two rewards calculation requests are sent out at slot `N` and`N+2`, since they
+both cross the epoch boundary and are on different forks. But the second
+computation is not necessary because they share the same parent node before the
+start of the epoch. The inputs, vote_accounts/stake_account, to the rewards
+calculation will be identical.
+
+To avoid this repeated computation, the signature of the computation requests
+is calculated.
+
+```
+calculation_signature = hash(epoch_number, [&stake_accounts], [&vote_accounts], [&stake_delegations])
+```
+
+Computation requests with same signature are discarded. In the above example,
+there are no stake/vote accounts changes between slot `N` and `N+2`, and the
+second computation request will be discarded.
+
+The reward computation phase lasts for `N` blocks. When reaching block height
+`N` after the start of the `reward calculation phase`, the bank will mark it
+the end the `reward calculation phase`.
+
+At this point, the reward computation result should be available. In case that
+the reward computation result is not available yet at the `N_th_` block-height,
+the bank runtime will halt and wait until the computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase, a set of sysvar accounts will be
+updated - `EpochRewardProof` and `RewardReserve`.
+
+`EpochRewardProof` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardProof` is an array of entries,
+```
+[(epoch, (reward_amount, root_hash))]
+```
+, where each entry corresponds to the reward for
+one particular epoch. The maximum depth of the history are the same as the
+depth of stake history, which is 512. The `root_hash` of the merkle tree for
```

Comment:
I think we should remove the merkle tree stuff entirely. If we need to add it in the future then we can deal with it then